### PR TITLE
WIP: update rust edition 2021

### PR DESCRIPTION
**Pull request by smrnjeet222**
Issue: https://github.com/project-serum/anchor/issues/1026
is this enough ?

### CODE REVIEW

**Comment by smrnjeet222** [comment]
changed:
[cli/src/template.rs](https://github.com/project-serum/anchor/pull/1030/files#diff-375957ff9bae6b1e9db964418670fa774c5a764ce430aa531a4a27a82a1ea8a5)
[lang/syn/src/codegen/accounts/to_account_metas.rs](https://github.com/project-serum/anchor/pull/1030/files#diff-375957ff9bae6b1e9db964418670fa774c5a764ce430aa531a4a27a82a1ea8a5)
[lang/syn/src/codegen/accounts/__cpi_client_accounts.rs](https://github.com/project-serum/anchor/pull/1030/files#diff-41be0a423037b29dca98a6090043d2732c131fb9e4d7a75b6097300abf1d262a)

File: lang/syn/src/codegen/accounts/__cpi_client_accounts.rs [hunk 1]
```diff
@@ -137,12 +137,12 @@ pub fn generate(accs: &AccountsStruct) -> proc_macro2::TokenStream {
...
             #(#re_exports)*
 
-            pub struct #name#generics {
+            pub struct #name #generics {
                 #(#account_struct_fields),*
             }
 
             #[automatically_derived]
-            impl#generics anchor_lang::ToAccountMetas for #name#generics {
+            impl #generics anchor_lang::ToAccountMetas for #name #generics {
```
Comment (smrnjeet222) [inline lang/syn/src/codegen/accounts/__cpi_client_accounts.rs:145]: image

Comment (smrnjeet222) [reply lang/syn/src/codegen/accounts/__cpi_client_accounts.rs:145]: not sure if this is correct

File: cli/src/template.rs [hunk 2]
```diff
@@ -48,11 +48,12 @@ export const IDL: {} = {};
 
 pub fn cargo_toml(name: &str) -> String {
     format!(
-        r#"[package]
+        r#"cargo-features = ["edition2021"]
```
Comment (fanatid) [inline cli/src/template.rs:51]: Is this can be a replacement of field `rust-version` in Cargo.toml?

Comment (smrnjeet222) [reply cli/src/template.rs:51]: yes, and it's either way necessary for edition=2021

**Comment by smrnjeet222** [comment]
can someone help me out, why Travis CI is failing!

File: cli/src/template.rs (hunk 2 above)
Comment (archseer) [reply cli/src/template.rs:51]: I don't think so, I'm only getting the above error when running an older cargo version. This feature should not be necessary.

Comment (Necmttn) [reply cli/src/template.rs:51]: at 
cargo 1.57.0 (b2e52d7ca 2021-10-21)
rustc 1.57.0 (f1edd0429 2021-11-29)
a.k.a latest version i'm still having the error above.

**Comment by luke7mm** [comment]
@smrnjeet222 Is this still unresolved? I was thinking about working on this too

**Pull request closed without merging**